In [1]:
# # Run this and then restart the kernel at the start of each session to install
# # 'teotil3' in development mode
# !pip install -e /home/jovyan/projects/teotil3/

In [2]:
import warnings

import matplotlib.pyplot as plt
import networkx as nx
import nivapy3 as nivapy
import numpy as np
import pandas as pd
import seaborn as sn
import teotil3 as teo
from sqlalchemy import text

warnings.simplefilter(action="ignore", category=FutureWarning)
sn.set_context("notebook")

# Estimating loads in unmonitored regions - 2023
# Using TEOTIL3

The standard workflow for OSPAR reporting uses TEOTIL2. This year (i.e. the 2024 analysis reporting 2023 data) we will switch to using TEOTIL3, because NIBIO are no longer able to generate the required input data for TEOTIL2.

The most recent notebook generating OSPAR data using TEOTIL2 is [here](https://nbviewer.org/github/JamesSample/rid/blob/master/notebooks/prog_2021-25/2022/10_loads_unmonitored_regions_2022.ipynb) (reporting of the 2022 data, undertaken during 2023). **This notebook translates the workflow for TEOTIL3**.

In [3]:
# Connect to db
engine = nivapy.da.connect_postgis()

Connection successful.


## 1. Read model results

In [4]:
# Year of interest
year = 2023

In [5]:
mod_csv = f"/home/jovyan/shared/common/teotil3/evaluation/teo3_results_nve{year+1}_2013-{year}_agri-annual-loss.csv"
mod_df = pd.read_csv(mod_csv).query("year == @year")
mod_df.head()

regine regine_down  accum_agriculture-background_din_kg  \
242020    001.10        001.                             0.000000   
242021  001.1A2B    001.1A2A                           134.896662   
242022  001.1A4D    001.1A4C                            16.342464   
242023    001.1M      001.1L                             0.000000   
242024    001.21        001.                             0.000000   

        accum_agriculture-background_ss_kg  \
242020                            0.000000   
242021                            0.989844   
242022                            0.382333   
242023                            0.000000   
242024                            0.000000   

        accum_agriculture-background_tdp_kg  \
242020                             0.000000   
242021                             4.190777   
242022                             0.143554   
242023                             0.000000   
242024                             0.000000   

        accum_agriculture-background_toc_kg  \
242020                             0.000000   
242021                          7769.727724   
242022                           569.571004   
242023                             0.000000   
242024                             0.000000   

        accum_agriculture-background_ton_kg  \
242020                             0.000000   
242021                            74.879508   
242022                             6.909292   
242023                             0.000000   
242024                             0.000000   

        accum_agriculture-background_totn_kg  \
242020                              0.000000   
242021                            209.776170   
242022                             23.251756   
242023                              0.000000   
242024                              0.000000   

        accum_agriculture-background_totp_kg  \
242020                              0.000000   
242021                              4.421938   
242022                              0.183389   
242023                              0.000000   
242024                              0.000000   

        accum_agriculture-background_tpp_kg  ...  local_urban_tpp_kg  \
242020                             0.000000  ...                 0.0   
242021                             0.231160  ...                 6.8   
242022                             0.039835  ...                 0.0   
242023                             0.000000  ...                 0.0   
242024                             0.000000  ...                 0.0   

        local_wood_din_kg  local_wood_ss_kg  local_wood_tdp_kg  \
242020               44.2             849.2                0.8   
242021             1614.9           29798.4               29.3   
242022              290.7            5272.5                5.3   
242023              580.1           11104.0               10.7   
242024                6.1             117.0                0.1   

        local_wood_toc_kg  local_wood_ton_kg  local_wood_totn_kg  \
242020            12180.8              228.2               272.4   
242021           470665.6             8746.0             10360.9   
242022            85814.9             1588.9              1879.6   
242023           174941.2             3236.4              3816.5   
242024             1699.0               31.7                37.8   

        local_wood_totp_kg  local_wood_tpp_kg  year  
242020                 6.7                5.9  2023  
242021               257.3              228.0  2023  
242022                46.9               41.6  2023  
242023                95.9               85.2  2023  
242024                 0.9                0.8  2023  

[5 rows x 164 columns]

In [6]:
# Save version with main catchments only
main_list = ["%03d." % i for i in range(1, 316)]
main_mod_df = mod_df.query("regine in @main_list").copy()
main_mod_df.sort_values("regine", inplace=True)
main_csv = f"/home/jovyan/shared/common/JES/teotil3_data/results/unmon_loads/teotil3_results_{year}_main_catchs.csv"
main_mod_df.to_csv(main_csv, index=False, encoding="utf-8")

## 4. Explore results

###  4.1. Identify areas with monitoring data

Where observations are available, we want to use them in preference to the model output. This means identifying all the catchments with observed data and substracting the model results for these locations. This is more complicated than it appears, because a small number of observed catchments are upstream of others, so subtracting all the loads for the 155 monitored catchments involves "double accounting", which we want to avoid. The first step is therefore to identify the downstream-most nodes for the monitored areas i.e. for the cases where one monitoring site/catchment is upstream of another, we just want the downstream node.

In [7]:
# Get regine polygons
reg_gdf = teo.io.get_regine_geodataframe(engine, year)

# Determine hydrological connectivity
reg_gdf = teo.io.assign_regine_hierarchy(
    reg_gdf,
    regine_col="regine",
    regine_down_col="regine_down",
    order_coastal=False,
    nan_to_vass=True,
    land_to_vass=True,
    add_offshore=True,
).dropna(subset="geometry")

# Read RID station data
in_xlsx = r"/home/jovyan/shared/common/JES/teotil2_data/RID_Sites_List_2017-2020.xlsx"
stn_df = pd.read_excel(in_xlsx, sheet_name="RID_All")

# Assign stations to regines
stn_df = nivapy.spatial.identify_point_in_polygon(
    stn_df,
    reg_gdf,
    pt_col="station_id",
    poly_col="regine",
    lat_col="lat",
    lon_col="lon",
)
assert stn_df["regine"].isna().sum() == 0

# Get just cols of interest and drop duplicates
# (some sites are in the same regine)
stn_df = stn_df[["ospar_region", "regine"]].drop_duplicates()

# Get regine IDs with observed data
obs_nds = set(stn_df["regine"].values)

# Build catchment network
g = teo.model.build_graph(
    reg_gdf[["regine", "regine_down"]], id_col="regine", next_down_col="regine_down"
)

# Get nodes upstream of each monitored site
nd_set = set()
for nd in obs_nds:
    nds = nx.dfs_tree(g.reverse(), nd).nodes()
    nd_set.update(nds)

# Get subgraph and ordered node list for all upstream nodes
g = g.subgraph(nd_set).copy()
nd_list = list(nx.topological_sort(g))

# Get downstream-most nodes in the subgraph
ds_nds = []
for nd in g:
    # If no downstream nodes
    if g.out_degree(nd) == 0:
        # Node is of interest
        ds_nds.append(nd)

# Get just the downstream catchments
stn_df = stn_df[stn_df["regine"].isin(ds_nds)]

100.00 % of regines assigned.


### 4.2. Sum model results for monitored locations

In [8]:
# Join accumulated outputs to stns of interest => modelled results for monitored areas
mod_mon_df = pd.merge(stn_df, mod_df, how="left", on="regine")

# Groupby OSPAR region
mod_mon_df = mod_mon_df.groupby("ospar_region").sum()

# Get just accum cols
pars = ["totn", "ton", "din", "totp", "tdp", "tpp", "toc", "ss"]
cols = ["accum_q_m3/s"] + [
    i
    for i in mod_mon_df.columns
    if (i.split("_")[0] == "accum") and (i.split("_")[-2] in pars)
]
mod_mon_df = mod_mon_df[cols]

# kg => tonnes
for col in mod_mon_df.columns:
    if col.endswith("_kg"):
        new_col = col.replace("_kg", "_tonnes")
        mod_mon_df[new_col] = mod_mon_df[col] / 1000
        del mod_mon_df[col]

mod_mon_df

accum_q_m3/s  accum_agriculture-background_din_tonnes  \
ospar_region                                                                 
LOFOTEN-BARENTS SEA    988.431487                                21.808215   
NORTH SEA             1568.938802                               348.159079   
NORWEGIAN SEA2        1809.086871                               280.757990   
SKAGERAK              2722.289810                              1310.122412   

                     accum_agriculture-background_ss_tonnes  \
ospar_region                                                  
LOFOTEN-BARENTS SEA                              153.957717   
NORTH SEA                                       2750.500664   
NORWEGIAN SEA2                                  3164.129624   
SKAGERAK                                        7492.696299   

                     accum_agriculture-background_tdp_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                                 0.197331   
NORTH SEA                                           3.808399   
NORWEGIAN SEA2                                      3.163844   
SKAGERAK                                           12.354076   

                     accum_agriculture-background_toc_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                               251.294244   
NORTH SEA                                        4366.209212   
NORWEGIAN SEA2                                   4385.234409   
SKAGERAK                                        24903.244390   

                     accum_agriculture-background_ton_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                                 6.153001   
NORTH SEA                                         131.254823   
NORWEGIAN SEA2                                     81.638631   
SKAGERAK                                          396.509395   

                     accum_agriculture-background_totn_tonnes  \
ospar_region                                                    
LOFOTEN-BARENTS SEA                                 27.961216   
NORTH SEA                                          479.413902   
NORWEGIAN SEA2                                     362.396621   
SKAGERAK                                          1706.631808   

                     accum_agriculture-background_totp_tonnes  \
ospar_region                                                    
LOFOTEN-BARENTS SEA                                  0.574467   
NORTH SEA                                            9.135189   
NORWEGIAN SEA2                                       8.579579   
SKAGERAK                                            26.206189   

                     accum_agriculture-background_tpp_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                                 0.377136   
NORTH SEA                                           5.326790   
NORWEGIAN SEA2                                      5.415735   
SKAGERAK                                           13.852114   

                     accum_agriculture_din_tonnes  ...  \
ospar_region                                       ...   
LOFOTEN-BARENTS SEA                    196.273932  ...   
NORTH SEA                             3136.348499  ...   
NORWEGIAN SEA2                        2526.821914  ...   
SKAGERAK                             11791.101711  ...   

                     accum_urban_totp_tonnes  accum_urban_tpp_tonnes  \
ospar_region                                                           
LOFOTEN-BARENTS SEA                 3.129397                1.170179   
NORTH SEA                          24.417848                7.541944   
NORWEGIAN SEA2                     20.929470                7.664675   
SKAGERAK                           73.358462               18.095606   

                     accum_wood_din_tonnes  accum_wood_ss_tonnes  \
o

This table gives the **modelled** inputs to each OSPAR region **from catchments for which we have observed data**. We want to subtract these values from the overall modelled inputs to each region and substitute the observed data instead.

### 4.3. Aggregate model output according to OSPAR regions

This code sums the model output for all regions (i.e. including the monitored areas).

In [9]:
# Define OSPAR regions (ranges are inclusive)
os_dict = {
    "SKAGERAK": (1, 23),
    "NORTH SEA": (24, 90),
    "NORWEGIAN SEA2": (91, 170),
    "LOFOTEN-BARENTS SEA": (171, 247),
}

# Container for results
df_list = []

# Loop over model output
for reg in os_dict.keys():
    min_id, max_id = os_dict[reg]

    vassoms = ["%03d." % i for i in range(min_id, max_id + 1)]

    # Get data for this region
    os_reg_df = mod_df.query("regine in @vassoms").copy()

    # Get just accum cols
    cols = ["accum_q_m3/s"] + [
        i
        for i in os_reg_df.columns
        if (i.split("_")[0] == "accum") and (i.split("_")[-2] in pars)
    ]
    os_reg_df = os_reg_df[cols]

    # Add region
    os_reg_df["ospar_region"] = reg

    # Add to output
    df_list.append(os_reg_df)

# Build df
os_df = pd.concat(df_list, axis=0)

# Aggregate
os_df = os_df.groupby("ospar_region").sum()

# kg => tonnes
for col in os_df.columns:
    if col.endswith("_kg"):
        new_col = col.replace("_kg", "_tonnes")
        os_df[new_col] = os_df[col] / 1000
        del os_df[col]

os_df.head()

accum_q_m3/s  accum_agriculture-background_din_tonnes  \
ospar_region                                                                 
LOFOTEN-BARENTS SEA   2542.634654                               200.809696   
NORTH SEA             3349.471533                              1073.010303   
NORWEGIAN SEA2        3859.262123                              1052.432324   
SKAGERAK              2905.710908                              1561.233117   

                     accum_agriculture-background_ss_tonnes  \
ospar_region                                                  
LOFOTEN-BARENTS SEA                             2937.794851   
NORTH SEA                                      13513.451262   
NORWEGIAN SEA2                                 15745.335807   
SKAGERAK                                       10286.401096   

                     accum_agriculture-background_tdp_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                                 2.447637   
NORTH SEA                                          12.292723   
NORWEGIAN SEA2                                     12.655361   
SKAGERAK                                           14.721484   

                     accum_agriculture-background_toc_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                              2356.263175   
NORTH SEA                                       13886.168939   
NORWEGIAN SEA2                                  19490.387195   
SKAGERAK                                        30637.654365   

                     accum_agriculture-background_ton_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                                63.303271   
NORTH SEA                                         403.497179   
NORWEGIAN SEA2                                    312.115555   
SKAGERAK                                          463.923183   

                     accum_agriculture-background_totn_tonnes  \
ospar_region                                                    
LOFOTEN-BARENTS SEA                                264.112967   
NORTH SEA                                         1476.507481   
NORWEGIAN SEA2                                    1364.547879   
SKAGERAK                                          2025.156299   

                     accum_agriculture-background_totp_tonnes  \
ospar_region                                                    
LOFOTEN-BARENTS SEA                                  7.532830   
NORTH SEA                                           34.025276   
NORWEGIAN SEA2                                      36.800403   
SKAGERAK                                            33.018535   

                     accum_agriculture-background_tpp_tonnes  \
ospar_region                                                   
LOFOTEN-BARENTS SEA                                 5.085193   
NORTH SEA                                          21.732553   
NORWEGIAN SEA2                                     24.145042   
SKAGERAK                                           18.297051   

                     accum_agriculture_din_tonnes  ...  \
ospar_region                                       ...   
LOFOTEN-BARENTS SEA                   1808.943236  ...   
NORTH SEA                             9671.958146  ...   
NORWEGIAN SEA2                        9472.294332  ...   
SKAGERAK                             14051.098049  ...   

                     accum_urban_totp_tonnes  accum_urban_tpp_tonnes  \
ospar_region                                                           
LOFOTEN-BARENTS SEA                41.900725               16.326600   
NORTH SEA                         170.653199               63.226042   
NORWEGIAN SEA2                    122.467032               46.638014   
SKAGERAK                          138.412102               42.235119   

                     accum_wood_din_tonnes  accum_wood_ss_tonnes  \
o

### 4.4. Estimate loads in unmonitored areas

We can now calculate the unmonitored component by subtracting the values modelled upstream of monitoring stations (`mod_mon_df`) from the overall modelled inputs to each OSPAR region (`os_df`).

In [10]:
# Calc unmonitored loads
assert (mod_mon_df.columns == os_df.columns).all()
assert (mod_mon_df.index == os_df.index).all()
unmon_df = (os_df - mod_mon_df).copy()

# Save
out_csv = f"/home/jovyan/shared/common/JES/teotil3_data/results/unmon_loads/teotil3_raw_unmonitored_loads_{year}.csv"
unmon_df.to_csv(out_csv, encoding="utf-8", index_label="ospar_region")

### 4.5. Aggregate to required quantities for OSPAR reporting

In [11]:
# Sources considered in OSPAR template
srcs = ["wastewater", "industry", "aquaculture", "diffuse"]

# Some variables with values of zero are not included in the model output.
# Add columns of zeros for completeness
unmon_df["accum_spredt_ss_tonnes"] = 0
unmon_df["accum_aquaculture_ss_tonnes"] = 0

# Aggregate
unmon_df["flow_1000m3pday"] = (
    unmon_df["accum_q_m3/s"] * 60 * 60 * 24 / 1000.0
)  # 1000s m3/day
for par in pars:
    # Lake deposition is zero for most pars, but not explicitly included in the
    # model output. Add here for completeness
    lake_par_col = f"accum_lake_{par}_tonnes"
    if lake_par_col not in unmon_df.columns:
        unmon_df[lake_par_col] = 0

    # Combine model outputs
    unmon_df[f"wastewater_{par}"] = (
        unmon_df[f"accum_large-wastewater_{par}_tonnes"]
        + unmon_df[f"accum_spredt_{par}_tonnes"]
    )
    unmon_df[f"industry_{par}"] = unmon_df[f"accum_industry_{par}_tonnes"]
    unmon_df[f"aquaculture_{par}"] = unmon_df[f"accum_aquaculture_{par}_tonnes"]
    unmon_df[f"diffuse_{par}"] = (
        unmon_df[f"accum_agriculture-background_{par}_tonnes"]
        + unmon_df[f"accum_agriculture_{par}_tonnes"]
        + unmon_df[f"accum_lake_{par}_tonnes"]
        + unmon_df[f"accum_upland_{par}_tonnes"]
        + unmon_df[f"accum_urban_{par}_tonnes"]
        + unmon_df[f"accum_wood_{par}_tonnes"]
    )

cols = ["flow_1000m3pday"] + [f"{src}_{par}" for src in srcs for par in pars]
unmon_df = unmon_df[cols].copy()

# Total for Norway
unmon_df.loc["NORWAY"] = unmon_df.sum(axis=0)

## 5. Other N and P species

Originally, TEOTIL2 only simulated TOTN and TOTP, and the table below was used to divide these into relevant subfractions for OSPAR reporting.

|   Source    | Phosphate | Nitrate | Ammonium |
|:-----------:|:---------:|:-------:|:--------:|
|    Sewage   |     0.600 |   0.050 |    0.750 |
|   Industry  |     0.600 |   0.050 |    0.750 |
| Aquaculture |     0.690 |   0.110 |    0.800 |
|   Diffuse   |     0.246 |   0.625 |    0.055 |

In addition, discharges from point sources of TOC, SS and metals were included where reported, and it was assumed that all these discharges reach the coast. 

TEOTIL3 explicitly includes DIN, TDP, TOC and SS. I will therefore use simulated values from TEOTIL3 where possible, and will only use directly reported point discharges for metals. To match the OSPAR parameters, I will assume that:

 * TDP in TEOTIL3 is approximately equal to PO4.
   
 * DIN in TEOTIL3 can be split into NO3 and NH4 using the same proportions as defined previously for the TEOTIL2 workflow (see table above). For example, I will assume that $\frac{5}{80} \times DIN = NO3$ and $\frac{75}{80} \times DIN = NH4$ etc.

In [12]:
# Factors for subdividing DIN
din_conv_dict = {
    ("wastewater", "no3"): 5 / 80,
    ("wastewater", "nh4"): 75 / 80,
    ("industry", "no3"): 5 / 80,
    ("industry", "nh4"): 75 / 80,
    ("aquaculture", "no3"): 11 / 91,
    ("aquaculture", "nh4"): 80 / 91,
    ("diffuse", "no3"): 625 / 680,
    ("diffuse", "nh4"): 55 / 680,
}

for col in unmon_df.drop(columns="flow_1000m3pday").columns:
    src, par = col.split("_")
    if par == "tdp":
        # Assume TDP = PO4
        unmon_df.rename(columns={col: f"{src}_po4"}, inplace=True)
    if par == "din":
        # Assume DIN = NO3 + NH4 and keep same proportions per source as previously
        unmon_df[f"{src}_no3"] = din_conv_dict[(src, "no3")] * unmon_df[col]
        unmon_df[f"{src}_nh4"] = din_conv_dict[(src, "nh4")] * unmon_df[col]
        del unmon_df[col]

unmon_df.round().astype(int).T

ospar_region      LOFOTEN-BARENTS SEA  NORTH SEA  NORWEGIAN SEA2  SKAGERAK  \
flow_1000m3pday                134283     153838          177135     15848   
wastewater_totn                  1378       4503            2935      5136   
wastewater_ton                    435        941             864      1062   
wastewater_totp                   168        482             384        96   
wastewater_po4                     78        213             172        33   
wastewater_tpp                     90        269             213        63   
wastewater_toc                   2581       8004            6899      5613   
wastewater_ss                    2284       6045            5265      2927   
industry_totn                      74        435             650       845   
industry_ton                       74        348             326       132   
industry_totp                       8         90              54        17   
industry_po4                        4         45              27         8   
industry_tpp                        4         45              27         8   
industry_toc                      422       2065            2657      4797   
industry_ss                    171393       7327          193604       432   
aquaculture_totn                18424      18935           29902        21   
aquaculture_ton                  5711       5870            9270         6   
aquaculture_totp                 3208       3297            5207         4   
aquaculture_po4                   930        956            1510         1   
aquaculture_tpp                  2278       2341            3697         3   
aquaculture_toc                 88727      91190          144006        99   
aquaculture_ss                      0          0               0         0   
diffuse_totn                     8317      20066           19611      5582   
diffuse_ton                      5421       8464            9574      2024   
diffuse_totp                      226        543             548       195   
diffuse_po4                       109        279             248        76   
diffuse_tpp                       116        264             300       119   
diffuse_toc                    105204     170845          261800     49716   
diffuse_ss                     224195     135380          215269     62837   
wastewater_no3                     59        223             129       255   
wastewater_nh4                    884       3340            1942      3820   
industry_no3                        0          5              20        45   
industry_nh4                        0         82             304       668   
aquaculture_no3                  1537       1579            2494         2   
aquaculture_nh4                 11176      11486           18139        13   
diffuse_no3                      2662      10664            9225      3270   
diffuse_nh4                       234        938             812       288   

ospar_region      NORWAY  
flow_1000m3pday   481104  
wastewater_totn    13953  
wastewater_ton      3301  
wastewater_totp     1130  
wastewater_po4       496  
wastewater_tpp       634  
wastewater_toc     23097  
wastewater_ss      16521  
industry_totn       2004  
industry_ton         880  
industry_totp        168  
industry_po4          84  
industry_tpp          84  
industry_toc        9941  
industry_ss       372755  
aquaculture_totn   67282  
aquaculture_ton    20858  
aquaculture_totp   11715  
aquaculture_po4     3397  
aquaculture_tpp     8318  
aquaculture_toc   324023  
aquaculture_ss         0  
diffuse_totn       53577  
diffuse_ton        25483  
diffuse_totp        1512  
diffuse_po4          712  
diffuse_tpp          800  
diffuse_toc       587564  
diffuse_ss        637681  
wastewater_no3       666  
wastewater_nh4      9986  
industry_no3          70  
industry_nh4        1054  
aquaculture_no3     5612  
aquaculture_nh4    40813  
diffuse_no3        25821  
diffuse_nh4         2272

## 6. Metals

All reported discharges of metals are assumed to reach the coast. We only want data for catchments that are not monitored i.e. for regine IDs **not** in the graph of monitored catchments created above (not in `nd_list`).

In [13]:
def assign_ospar_region(vassom):
    if vassom in range(1, 24):
        return "SKAGERAK"
    elif vassom in range(24, 91):
        return "NORTH SEA"
    elif vassom in range(91, 171):
        return "NORWEGIAN SEA2"
    elif vassom in range(171, 248):
        return "LOFOTEN-BARENTS SEA"
    else:
        return np.nan

In [14]:
sectors = ["aquaculture", "industry", "large wastewater"]
metal_pars = [
    "as_kg",
    "cd_kg",
    "cr_kg",
    "cu_kg",
    "hg_kg",
    "ni_kg",
    "pb_kg",
    "zn_kg",
]
df_list = []
for sector in sectors:
    # Get reported discharges
    sec_pt_df = teo.io.get_annual_point_data(
        engine,
        year,
        sector,
        par_list=metal_pars,
    )

    # Filter to only consider unmonitoried regions
    sec_pt_df = sec_pt_df.query("regine not in @nd_list")

    # Group to OSPAR region
    sec_pt_df["vassom"] = sec_pt_df["regine"].str[:3].astype(int)
    sec_pt_df["osp_reg"] = sec_pt_df["vassom"].apply(assign_ospar_region)
    sec_pt_df = sec_pt_df.groupby("osp_reg").sum(numeric_only=True)
    del sec_pt_df["vassom"]
    df_list.append(sec_pt_df)
pt_df = pd.concat(df_list, axis="columns")

# Rename
col_map = {
    col: col.replace("large-wastewater_", "wastewater_") for col in pt_df.columns
}
pt_df.rename(columns=col_map, inplace=True)

# kg => tonnes and remove units
for col in pt_df.columns:
    if col.endswith("_kg"):
        new_col = col.replace("_kg", "")
        pt_df[new_col] = pt_df[col] / 1000
        del pt_df[col]

# Total for Norway
pt_df.loc["NORWAY"] = pt_df.sum(axis=0)

pt_df.T

osp_reg         LOFOTEN-BARENTS SEA  NORTH SEA  NORWEGIAN SEA2  SKAGERAK  \
                                                                           
aquaculture_cu             102.3907   105.2336        166.1824    0.1147   
industry_as                  0.0002     1.7158          0.2429    0.2131   
industry_cd                  0.0000     0.0316          0.0236    0.0115   
industry_cr                  0.0000     0.3243          0.2219    0.5171   
industry_cu                  0.0003     1.3483          0.9166    2.7785   
industry_hg                  0.0001     0.0036          0.0022    0.0027   
industry_ni                  0.0007     3.3297          0.2757    1.2442   
industry_pb                  0.0000     0.3266          0.0818    0.0852   
industry_zn                  0.6087     3.9162          2.7645    2.8460   
wastewater_as                0.0055     0.0902          0.0503    0.1370   
wastewater_cd                0.0005     0.0030          0.0018    0.0077   
wastewater_cr                0.0182     0.0871          0.1561    0.1692   
wastewater_cu                0.2185     0.9933          1.3661    2.8945   
wastewater_hg                0.0002     0.0005          0.0004    0.0019   
wastewater_ni                0.0439     0.3627          0.1667    1.3829   
wastewater_pb                0.0070     0.0829          0.0500    0.1550   
wastewater_zn                0.4721     3.2923          2.2606    7.3464   

osp_reg           NORWAY  
                          
aquaculture_cu  373.9214  
industry_as       2.1720  
industry_cd       0.0667  
industry_cr       1.0633  
industry_cu       5.0437  
industry_hg       0.0086  
industry_ni       4.8503  
industry_pb       0.4936  
industry_zn      10.1354  
wastewater_as     0.2830  
wastewater_cd     0.0130  
wastewater_cr     0.4306  
wastewater_cu     5.4724  
wastewater_hg     0.0030  
wastewater_ni     1.9562  
wastewater_pb     0.2949  
wastewater_zn    13.3714

In [15]:
# Join reported discharges of metals to model results
unmon_df = unmon_df.join(pt_df)
unmon_df.round(0).astype(int).T

ospar_region      LOFOTEN-BARENTS SEA  NORTH SEA  NORWEGIAN SEA2  SKAGERAK  \
flow_1000m3pday                134283     153838          177135     15848   
wastewater_totn                  1378       4503            2935      5136   
wastewater_ton                    435        941             864      1062   
wastewater_totp                   168        482             384        96   
wastewater_po4                     78        213             172        33   
wastewater_tpp                     90        269             213        63   
wastewater_toc                   2581       8004            6899      5613   
wastewater_ss                    2284       6045            5265      2927   
industry_totn                      74        435             650       845   
industry_ton                       74        348             326       132   
industry_totp                       8         90              54        17   
industry_po4                        4         45              27         8   
industry_tpp                        4         45              27         8   
industry_toc                      422       2065            2657      4797   
industry_ss                    171393       7327          193604       432   
aquaculture_totn                18424      18935           29902        21   
aquaculture_ton                  5711       5870            9270         6   
aquaculture_totp                 3208       3297            5207         4   
aquaculture_po4                   930        956            1510         1   
aquaculture_tpp                  2278       2341            3697         3   
aquaculture_toc                 88727      91190          144006        99   
aquaculture_ss                      0          0               0         0   
diffuse_totn                     8317      20066           19611      5582   
diffuse_ton                      5421       8464            9574      2024   
diffuse_totp                      226        543             548       195   
diffuse_po4                       109        279             248        76   
diffuse_tpp                       116        264             300       119   
diffuse_toc                    105204     170845          261800     49716   
diffuse_ss                     224195     135380          215269     62837   
wastewater_no3                     59        223             129       255   
wastewater_nh4                    884       3340            1942      3820   
industry_no3                        0          5              20        45   
industry_nh4                        0         82             304       668   
aquaculture_no3                  1537       1579            2494         2   
aquaculture_nh4                 11176      11486           18139        13   
diffuse_no3                      2662      10664            9225      3270   
diffuse_nh4                       234        938             812       288   
aquaculture_cu                    102        105             166         0   
industry_as                         0          2               0         0   
industry_cd                         0          0               0         0   
industry_cr                         0          0               0         1   
industry_cu                         0          1               1         3   
industry_hg                         0          0               0         0   
industry_ni                         0          3               0         1   
industry_pb                         0          0               0         0   
industry_zn                         1          4               3         3   
wastewater_as                       0          0               0         0   
wastewater_cd                       0          0               0         0   
wastewater_cr                       0          0               0         0   
wastewater_cu                       0          1               1         3   
wastewater_hg         

In [16]:
# Save
out_csv = f"/home/jovyan/shared/common/JES/teotil3_data/results/unmon_loads/teotil3_ospar_unmonitored_loads_{year}.csv"
unmon_df.to_csv(out_csv)

This dataset can be used to create Table 3 in the report and/or the OSPAR template.